library/alloc/src/string.rs

```diff
@@ -117,27 +117,84 @@ use crate::vec::Vec;
 ///
 /// # UTF-8
 ///
-/// `String`s are always valid UTF-8. This has a few implications, the first of
-/// which is that if you need a non-UTF-8 string, consider [`OsString`]. It is
-/// similar, but without the UTF-8 constraint. The second implication is that
-/// you cannot index into a `String`:
+/// `String`s are always valid UTF-8. If you need a non-UTF-8 string, consider
+/// [`OsString`]. It is similar, but without the UTF-8 constraint. Because UTF-8
+/// is a variable width encoding, `String`s can be smaller than an array of
+/// the same `chars`:
+///
+/// ```
+/// use std::mem;
+///
+/// // `s` is ASCII which represents each `char` as one byte
+/// let s = "hello";
+/// assert_eq!(s.len(), 5);
+///
+/// // A `char` array with the same contents would be longer because
+/// // every `char` is four bytes
+/// let s = ['h', 'e', 'l', 'l', 'o'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
+/// assert_eq!(size, 20);
+///
+/// // However, for non-ASCII strings, the difference will be smaller
+/// // and sometimes they are the same
+/// let s = "💖💖💖💖💖";
+/// assert_eq!(s.len(), 20);
+///
+/// let s = ['💖', '💖', '💖', '💖', '💖'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
+/// assert_eq!(size, 20);
+/// ```
+///
+/// This raises an interesting question as to how `s[i]` should work.
+/// What should `i` be here - a byte index or a `char` index? Because
```

Comment:
I think it might be worth noting here that there are other options (e.g., grapheme indexes), but they are similarly not constant time -- really, the problem here is that because these are not constant time operations, you are unlikely to actually want just the nth grapheme or char -- you probably want to iterate, so we have chars() and there's graphemes() somewhere in the ecosystem.

Reply:
Yeah I had intentionally not mentioned graphemes (and I didn't love them in the previous example) because nothing else in `std::` had anything to do with graphemes (to my knowledge!). But I can mention it with the reasoning above!

Reply:
> you are unlikely to actually want just the nth grapheme or char -- you probably want to iterate

I'm also not 100% sold on this reasoning (That we didn't provide an index operation because you probably wanted to iterate anyway because ... what if the user _doesn't_ want to iterate? Seems like a weird assumption to me :thinking:) but I think

> really, the problem here is that because these are not constant time operations

is incontrovertible. Is it okay if I just focus on that or do we think mentioning the other thing is important? I might just be missing the value! :sweat_smile:

Reply:
I think not mentioning graphemes is probably ok (it might make sense to have some notes on that topic in char docs but probably not mentioning here is good).

The point is that iterators "obviously" expose non-constant time access via nth(...) so you essentially already have that API, and since it's typically not what you want we avoid making an easy to use but actually wrong API by implementing index with &u8 return type.

Reply:
Hopefully handled in a52ab6946e508fdfe4f8dc6313f9b7d9e0d77b58